[FIX] core: calling read_group / name_search over xmlrpc

xmo-odoo opened the pull request:
Also non-browser jsonrpc (as it goes through a similar process): for
internal performance reasons, name_search and read_group have been
converted to a *lazy* name_get, so the "display name" is not
unnecessarily computed.

However this is an issue for the RPC endpoints (/xmlrpc and /jsonrpc)
as they have no support for `lazy` and thus tend to blow up and / or
do the wrong thing when trying to output a lazy:

* xmlrpc has no way to handle lazy at all and straight blows up
* jsonrpc falls back to `json_default` so they try to stringify the
  lazy, which might have worked except

*Problematically* both endpoints delegate the actual work to
`dispatch_rpc` which handles dispatching between various services and
ultimately creates a *new* cursor before calling model
methods (`object` service and `execute`/`execute_kw`).

This means by the time the result is serialized to be output, the
lazy's cursor has long been closed, and thus any access to an
unevaluated `lazy` errors out when trying to fetch the underlying
item.

This also means we can't just add a hook to serialize the lazy
in the xmlrpc marshaller, though we do have to do that. We *also* (for
both xmlrpc and jsonrpc) have to force evluation of lazy values before
our cursor is closed, meaning it has to be done right after the method
is invoked, iterating the entire response.

Related to task 2170343

rco-odoo commented on odoo/service/model.py:
```diff
@@ -165,7 +166,10 @@ def execute_cr(cr, uid, obj, method, *args, **kw):
-    return odoo.api.call_kw(recs, method, args, kw)
+    result = odoo.api.call_kw(recs, method, args, kw)
+    for l in iterate(result, lazy):
+        _0 = l._value # evaluate and cache while we still have a valid cursor
```
I would have put a comment above the `for` loop: "force evaluation of lazy values before closing the cursor".
I was wondering whether `l._value` might still contain lazy values, but that's not the case with `lazy_name_get`.
So that's fine.

rco-odoo commented on odoo/tools/__init__.py:
```diff
@@ -18,3 +19,17 @@
+    if isinstance(obj, lazy):
+        return obj._value
+    return ustr(obj)
+date_utils.json_default = json_default
```
I wonder whether this is still relevant: https://github.com/odoo/odoo/blob/2a2ff6eb5c746acec46ebc7fee382dfa78ec8c14/odoo/tools/func.py#L247-L255

xmo-odoo replied:
Might be for arbitrary / random json.dumps I guess? Unclear whether there's a risk those would get a lazy coming their way.

xmo-odoo replied:
Tried to remove it, was only used (nearly explicitly) in a test, so I updated the test: https://github.com/odoo/odoo/blob/307be80ec659c6c2d1e2371d2eb01d345fbb275f/odoo/addons/test_performance/tests/test_performance.py#L142-L144

odony replied:
> Tried to remove it, was only used (nearly explicitly) in a test, so I updated the test:

Why would you change this in a stable version? Was this part necessary to fix the bug?

rco-odoo commented on odoo/tools/misc.py:
```diff
@@ -26,7 +26,7 @@
 import zipfile
 from collections import defaultdict, Iterable, Mapping, MutableMapping, MutableSet, OrderedDict
-from itertools import islice, groupby as itergroupby, repeat
+from itertools import islice, groupby as itergroupby, repeat, chain
```
not used, in fact

xmo-odoo replied:
Yeah I was using chain.from_iterable in the original iterate but it didn't work right, forgot to remove the import when I changed the implementation.

Should probably rename iterate as well, though not sure what name to use there. iterate_struct? iterate_builtin?

xmo-odoo commented:
* lifted the comment above the for loop and expanded it a bit
* removed the json stuff by default, if there can be lazies in a value being hand-serialized there can probably be dates as well in which cases the `default` handler should probably be explicitly provided
* renamed iterator thing from the way too abstract `iterate` to the more concrete `traverse_containers`, ugly and not perfect but makes more sense given what it actually does... not quite ideal though as it doesn't really traverse [Containers](https://docs.python.org/3/library/collections.abc.html#collections.abc.Container), as Python's containers... contain things, they can't be traversed (that'd be iterable).

xmo-odoo commented:
@robodoo r+
